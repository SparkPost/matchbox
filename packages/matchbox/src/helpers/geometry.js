import React from 'react';
import { findDOMNode } from 'react-dom';
import { clamp } from './math';
import { debounce } from './event';

export function getWindowRect() {
  return {
    top: window.scrollY,
    left: window.scrollX,
    height: window.innerHeight,
    width: window.innerWidth
  };
}

export function getRectFor(node) {
  const rect = findDOMNode(node);

  if (!rect) {
    return {};
  }

  return rect.getBoundingClientRect();
}

/**
 * Provides a hook that returns window dimensions
 * @param  {number} [wait=100] Timer for debounced dimension calculation
 * @return {Shape}             Same results as `getWindowRect``
 */
export function useWindowSize(wait = 100) {
  const [size, setSize] = React.useState(getWindowRect());

  const handleResize = debounce(() => {
    setSize(getWindowRect());
  }, wait);

  React.useEffect(() => {
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return size;
}

/**
 * Gets preferred direction for the provided react component
 * @param  {React Node} node - a react component
 * @return {Shape}
 * Returns directional booleans for where the component should render. Eg:
 * top: true if component is in the bottom half of the screen
 * right: true if component in the left half of the screen
 */
export function getPreferredDirectionFor(node) {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  const rightOffset = windowRect.width - elementRect.right;
  const bottomOffset = windowRect.height - elementRect.bottom;

  return {
    top: bottomOffset < elementRect.top,
    left: rightOffset < elementRect.left,
    right: rightOffset >= elementRect.left,
    bottom: bottomOffset >= elementRect.top
  };
}


/**
 * Gets coordinates and dimensions in pixels for the provided react component
 * @param  {React Node} node
 * @return {Shape}
 */
export function getPositionFor(node, { fixed = false } = {}) {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  return {
    top: elementRect.top + (fixed ? 0 : windowRect.top),
    left: elementRect.left + (fixed ? 0 : windowRect.left),
    width: elementRect.width,
    height: elementRect.height
  };
}

/**
 * Linearly interpolates and clamps between two values
 * @param  {number} min
 * @param  {number} max
 * @param  {number} n
 * @return {number}
 * @example
 *   lerp(10, 20, 0.5)
 *   > 15
 */
export function lerp(min, max, n) {
  const value = (max - min) * n + min;
  return clamp(value, min, max);
}
