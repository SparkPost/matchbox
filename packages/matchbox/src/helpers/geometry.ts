import React from 'react';
import { findDOMNode } from 'react-dom';
import { clamp } from './math';
import { debounce } from './event';
import { getWindow } from './window';

export type WindowRect = {
  top: number;
  left: number;
  height: number;
  width: number;
};

export function getWindowRect(): WindowRect {
  const environment = getWindow();
  return {
    top: environment.scrollY,
    left: environment.scrollX,
    height: environment.innerHeight,
    width: environment.innerWidth,
  };
}

export function getRectFor(node?: React.ReactInstance): ClientRect {
  // Need to update this soon since findDomNode is deprecated
  const rect = findDOMNode(node) as Element; //eslint-disable-line

  if (!rect) {
    return {} as ClientRect;
  }

  return rect.getBoundingClientRect();
}

/**
 * Provides a hook that returns window dimensions
 * @param  {number} [wait=100] Timer for debounced dimension calculation
 * @return {Shape}             Same results as `getWindowRect``
 */
export function useWindowSize(wait = 100): WindowRect {
  const environment = getWindow();
  const [size, setSize] = React.useState(getWindowRect());

  const handleResize = debounce(() => {
    setSize(getWindowRect());
  }, wait);

  React.useEffect(() => {
    environment.addEventListener('resize', handleResize);
    return () => {
      environment.removeEventListener('resize', handleResize);
    };
  }, []);

  return size;
}

/**
 * Gets preferred direction for the provided react component
 * @param  {React Node} node - a react component
 * @return {Shape}
 * Returns directional booleans for where the component should render. Eg:
 * top: true if component is in the bottom half of the screen
 * right: true if component in the left half of the screen
 */
export function getPreferredDirectionFor(node?: React.ReactInstance): {
  top: boolean;
  left: boolean;
  right: boolean;
  bottom: boolean;
} {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  const rightOffset = windowRect.width - elementRect.right;
  const bottomOffset = windowRect.height - elementRect.bottom;

  return {
    top: bottomOffset < elementRect.top,
    left: rightOffset < elementRect.left,
    right: rightOffset >= elementRect.left,
    bottom: bottomOffset >= elementRect.top,
  };
}

/**
 * Gets coordinates and dimensions in pixels for the provided react component
 * @param  {React Node} node
 * @return {Shape}
 */
export function getPositionFor(node?: React.ReactInstance, { fixed = false } = {}): WindowRect {
  const windowRect = getWindowRect();
  const elementRect = getRectFor(node);

  return {
    top: elementRect.top + (fixed ? 0 : windowRect.top),
    left: elementRect.left + (fixed ? 0 : windowRect.left),
    width: elementRect.width,
    height: elementRect.height,
  };
}

/**
 * Linearly interpolates and clamps between two values
 * @param  {number} min
 * @param  {number} max
 * @param  {number} n
 * @return {number}
 * @example
 *   lerp(10, 20, 0.5)
 *   > 15
 */
export function lerp(min: number, max: number, n: number): number {
  const value = (max - min) * n + min;
  return clamp(value, min, max);
}

/**
 * Rounds a number to the nearest baseline ceiling
 */
export function roundToBaseline(n: number, base = 4): number {
  return Math.ceil(n / base) * base;
}
