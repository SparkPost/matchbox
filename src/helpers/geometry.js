import { findDOMNode } from 'react-dom';

export function getWindowRect() {
  return {
    top: window.scrollY,
    left: window.scrollX,
    height: window.innerHeight,
    width: window.innerWidth
  };
}

export function getRectForNode(node) {
  const rect = findDOMNode(node);

  if (!rect) {
    return {};
  }

  return rect.getBoundingClientRect();
}
