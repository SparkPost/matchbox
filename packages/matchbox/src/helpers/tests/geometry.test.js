import React from 'react';
import { mount } from 'enzyme';
import { getWindowRect, getRectFor, getPreferredDirectionFor, getPositionFor, lerp, useWindowSize } from '../geometry';
import * as reactDomMock from 'react-dom';
jest.mock('react-dom');

function MockComponent() {
  const size = useWindowSize();
  return <div {...size} />;
}

global.scrollY = 10;
global.scrollX = 20;
global.innerHeight = 600;
global.innerWidth = 900;

describe('getWindowRect', () => {
  it('should return window dimensions', () => {
    expect(getWindowRect()).toMatchSnapshot();
  });
});

describe('getRectFor', () => {

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should return empty object with no node', () => {
    reactDomMock.findDOMNode.mockImplementationOnce(() => undefined);
    const result = getRectFor('test');
    expect(reactDomMock.findDOMNode).toHaveBeenCalledWith('test');
    expect(result).toMatchSnapshot();
  });

  it('should return node dimensions', () => {
    reactDomMock.findDOMNode.mockImplementationOnce(() => ({ getBoundingClientRect: jest.fn(() => 'dimensions') }));
    const result = getRectFor('test');
    expect(reactDomMock.findDOMNode).toHaveBeenCalledWith('test');
    expect(result).toMatchSnapshot();
  });
});

describe('useWindowSize', () => {
  it('should return window dimensions', () => {
    const wrapper = mount(<MockComponent/>);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('getPreferredDirectionFor', () => {
  beforeEach(() => {
    reactDomMock.findDOMNode.mockImplementation(() => ({ getBoundingClientRect: jest.fn(() => ({ top: 600, left: 50, right: 80, bottom: 700 })) }));
  });

  it('should return direction', () => {
    expect(getPreferredDirectionFor('test')).toMatchSnapshot();
  });
});

describe('getPositionFor', () => {
  beforeEach(() => {
    reactDomMock.findDOMNode.mockImplementation(() => ({ getBoundingClientRect: jest.fn(() => ({ top: 600, left: 50, width: 20, height: 10 })) }));
  });

  it('should return direction', () => {
    expect(getPositionFor('test')).toMatchSnapshot();
  });

  it('should return direction if node is fixed', () => {
    expect(getPositionFor('test', { fixed: true })).toMatchSnapshot();
  });
});


describe('linear interpolation', () => {
  it('should correctly interpolate between two points', () => {
    expect(lerp(-10, 30, 0.5)).toBe(10);
  });

  it('should clamp lower bounds', () => {
    expect(lerp(-10, 30, -1.5)).toBe(-10);
  });

  it('should clamp upper bounds', () => {
    expect(lerp(-10, 30, 1.5)).toBe(30);
  });
});
