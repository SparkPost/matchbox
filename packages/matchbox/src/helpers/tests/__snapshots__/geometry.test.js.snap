// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPositionFor should return direction 1`] = `
Object {
  "height": 10,
  "left": 70,
  "top": 610,
  "width": 20,
}
`;

exports[`getPositionFor should return direction if node is fixed 1`] = `
Object {
  "height": 10,
  "left": 50,
  "top": 600,
  "width": 20,
}
`;

exports[`getPreferredDirectionFor should return direction 1`] = `
Object {
  "bottom": false,
  "left": false,
  "right": true,
  "top": true,
}
`;

exports[`getRectFor should return empty object with no node 1`] = `Object {}`;

exports[`getRectFor should return node dimensions 1`] = `"dimensions"`;

exports[`getWindowRect should return window dimensions 1`] = `
Object {
  "height": 600,
  "left": 20,
  "top": 10,
  "width": 900,
}
`;

exports[`useWindowSize should return window dimensions 1`] = `
<MockComponent>
  <div
    height={600}
    left={20}
    top={10}
    width={900}
  />
</MockComponent>
`;
